#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
#
# Author: Tracy Chew
# tracy.chew@sydney.edu.au
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

#PBS -P bt64
#PBS -N tpmtranscript_matrix
#PBS -l walltime=05:00:00,ncpus=1,mem=96GB,wd
#PBS -q express
#PBS -W umask=022
#PBS -l storage=scratch/bt64
#PBS -l wd
#PBS -o ./Logs/tpmcalculator_matrix.o
#PBS -e ./Logs/tpmcalculator_matrix.e

set -e

# Set the PBS_O_WORKDIR to directory that the scripts are being queued from
PBS_O_WORKDIR=/scratch/bt64/nn8573/etql/ceph/RNASeq-DE/Scripts
cd "${PBS_O_WORKDIR}"

# Replace with path to your TPMCalculator output
tpmdir=../cohort_TPMCalculator

# Create TPM transcript count matrix
perl ./tpmcalculator_transcript_make_matrix.pl $tpmdir

# Create TPM gene count matrix
perl ./tpmcalculator_make_matrix.pl $tpmdir
